File: deployment/build-scripts/build-tenant-site.js
#!/usr/bin/env node

/**
 * Tenant Site Builder
 * 
 * This script builds a customized tenant site from the template by:
 * 1. Cloning the template
 * 2. Injecting tenant-specific configuration
 * 3. Building the static site
 * 4. Preparing for deployment
 */

import fs from 'fs/promises';
import path from 'path';
import { execSync } from 'child_process';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

class TenantSiteBuilder {
  constructor() {
    this.templateDir = path.join(__dirname, '..', 'tenant-site-template');
    this.buildDir = path.join(__dirname, '..', 'builds');
    this.deploymentDir = path.join(__dirname, '..', 'deployments');
  }

  /**
   * Build a tenant site with the provided configuration
   */
  async buildSite(tenantConfig, buildId) {
    console.log(`🏗️  Building tenant site for: ${tenantConfig.subdomain}`);
    console.log(`📝 Build ID: ${buildId}`);

    try {
      // Create build directory
      const siteBuildDir = path.join(this.buildDir, buildId);
      await fs.mkdir(siteBuildDir, { recursive: true });

      // Copy template to build directory
      console.log('📂 Copying template files...');
      await this.copyTemplate(siteBuildDir);

      // Inject tenant configuration
      console.log('⚙️  Injecting tenant configuration...');
      await this.injectTenantConfig(siteBuildDir, tenantConfig);

      // Install dependencies
      console.log('📦 Installing dependencies...');
      execSync('npm install', { 
        cwd: siteBuildDir, 
        stdio: 'inherit',
        env: { ...process.env, NODE_ENV: 'production' }
      });

      // Build the site
      console.log('🔨 Building site...');
      execSync('npm run build', { 
        cwd: siteBuildDir, 
        stdio: 'inherit',
        env: { 
          ...process.env, 
          NODE_ENV: 'production',
          VITE_TENANT_CONFIG: JSON.stringify(tenantConfig)
        }
      });

      // Prepare deployment package
      const deploymentPackage = await this.prepareDeploymentPackage(siteBuildDir, buildId, tenantConfig);

      console.log('✅ Site built successfully!');
      console.log(`📦 Deployment package: ${deploymentPackage.packagePath}`);

      return {
        success: true,
        buildId,
        buildPath: siteBuildDir,
        deploymentPackage
      };

    } catch (error) {
      console.error('❌ Build failed:', error.message);
      
      // Clean up failed build
      try {
        const siteBuildDir = path.join(this.buildDir, buildId);
        await fs.rm(siteBuildDir, { recursive: true, force: true });
      } catch (cleanupError) {
        console.error('Failed to clean up build directory:', cleanupError.message);
      }

      return {
        success: false,
        buildId,
        error: error.message
      };
    }
  }

  /**
   * Copy template directory to build location
   */
  async copyTemplate(targetDir) {
    const copyRecursive = async (src, dest) => {
      const stats = await fs.stat(src);
      
      if (stats.isDirectory()) {
        await fs.mkdir(dest, { recursive: true });
        const entries = await fs.readdir(src);
        
        for (const entry of entries) {
          // Skip node_modules and other build artifacts
          if (['node_modules', '.git', 'dist', '.cache'].includes(entry)) {
            continue;
          }
          
          await copyRecursive(
            path.join(src, entry), 
            path.join(dest, entry)
          );
        }
      } else {
        await fs.copyFile(src, dest);
      }
    };

    await copyRecursive(this.templateDir, targetDir);
  }

  /**
   * Inject tenant configuration into the build
   */
  async injectTenantConfig(buildDir, config) {
    // Create tenant-specific configuration files
    const configPath = path.join(buildDir, 'src', 'config', 'injected-config.ts');
    const configContent = `
// Auto-generated tenant configuration
// This file is created during the build process

export const TENANT_CONFIG = ${JSON.stringify(config, null, 2)};

// Make config globally available
declare global {
  interface Window {
    __TENANT_CONFIG__: typeof TENANT_CONFIG;
  }
}

window.__TENANT_CONFIG__ = TENANT_CONFIG;
`;

    await fs.writeFile(configPath, configContent);

    // Update the main config file to use injected config
    const mainConfigPath = path.join(buildDir, 'src', 'config', 'tenant-config.ts');
    let mainConfigContent = await fs.readFile(mainConfigPath, 'utf-8');
    
    // Add import for injected config
    mainConfigContent = `import { TENANT_CONFIG } from './injected-config';\n${mainConfigContent}`;
    
    // Replace the loadTenantConfig function
    mainConfigContent = mainConfigContent.replace(
      /const loadTenantConfig = \(\)[\s\S]*?};/,
      `const loadTenantConfig = (): TenantConfig => {
  return TENANT_CONFIG;
};`
    );

    await fs.writeFile(mainConfigPath, mainConfigContent);

    // Create environment file
    const envPath = path.join(buildDir, '.env.production');
    const envContent = `
VITE_API_BASE_URL=https://api.base2ml.com
VITE_TENANT_ID=${config.tenantId}
VITE_SUBDOMAIN=${config.subdomain}
`;
    
    await fs.writeFile(envPath, envContent);

    // Update index.html with tenant-specific meta tags
    const indexPath = path.join(buildDir, 'index.html');
    let indexContent = await fs.readFile(indexPath, 'utf-8');
    
    // Inject meta tags
    const metaTags = `
    <title>${config.siteSettings.title}</title>
    <meta name="description" content="${config.siteSettings.description}" />
    <meta property="og:title" content="${config.siteSettings.title}" />
    <meta property="og:description" content="${config.siteSettings.description}" />
    <meta name="theme-color" content="${config.siteSettings.theme.primaryColor}" />
  `;
    
    indexContent = indexContent.replace('<title>Vite + React + TS</title>', metaTags);
    
    await fs.writeFile(indexPath, indexContent);
  }

  /**
   * Prepare deployment package
   */
  async prepareDeploymentPackage(buildDir, buildId, config) {
    const deploymentDir = path.join(this.deploymentDir, buildId);
    await fs.mkdir(deploymentDir, { recursive: true });

    // Copy built files
    const distDir = path.join(buildDir, 'dist');
    const packageDistDir = path.join(deploymentDir, 'dist');
    
    await this.copyDirectory(distDir, packageDistDir);

    // Create deployment manifest
    const manifest = {
      buildId,
      tenantId: config.tenantId,
      subdomain: config.subdomain,
      builtAt: new Date().toISOString(),
      config: {
        title: config.siteSettings.title,
        description: config.siteSettings.description,
        parentInfo: config.parentInfo
      },
      files: await this.getFileList(packageDistDir)
    };

    const manifestPath = path.join(deploymentDir, 'deployment-manifest.json');
    await fs.writeFile(manifestPath, JSON.stringify(manifest, null, 2));

    // Create deployment script
    const deployScript = `#!/bin/bash
set -e

echo "🚀 Deploying ${config.subdomain}.base2ml.com"
echo "📦 Build ID: ${buildId}"

# Sync files to S3 (or other static hosting)
if [ -n "$AWS_S3_BUCKET" ]; then
    aws s3 sync ./dist s3://$AWS_S3_BUCKET/${config.subdomain}/ --delete
    echo "✅ Files synced to S3"
else
    echo "⚠️  AWS_S3_BUCKET not set, skipping S3 sync"
fi

# Invalidate CloudFront cache
if [ -n "$AWS_CLOUDFRONT_DISTRIBUTION" ]; then
    aws cloudfront create-invalidation \\
        --distribution-id $AWS_CLOUDFRONT_DISTRIBUTION \\
        --paths "/${config.subdomain}/*"
    echo "✅ CloudFront cache invalidated"
else
    echo "⚠️  AWS_CLOUDFRONT_DISTRIBUTION not set, skipping cache invalidation"
fi

echo "🎉 Deployment complete!"
echo "🌐 Site available at: https://${config.subdomain}.base2ml.com"
`;

    const deployScriptPath = path.join(deploymentDir, 'deploy.sh');
    await fs.writeFile(deployScriptPath, deployScript);
    await fs.chmod(deployScriptPath, '755');

    return {
      packagePath: deploymentDir,
      manifestPath,
      deployScriptPath,
      manifest
    };
  }

  /**
   * Copy directory recursively
   */
  async copyDirectory(src, dest) {
    await fs.mkdir(dest, { recursive: true });
    const entries = await fs.readdir(src, { withFileTypes: true });

    for (const entry of entries) {
      const srcPath = path.join(src, entry.name);
      const destPath = path.join(dest, entry.name);

      if (entry.isDirectory()) {
        await this.copyDirectory(srcPath, destPath);
      } else {
        await fs.copyFile(srcPath, destPath);
      }
    }
  }

  /**
   * Get list of files in directory
   */
  async getFileList(dir, prefix = '') {
    const files = [];
    const entries = await fs.readdir(dir, { withFileTypes: true });

    for (const entry of entries) {
      const relativePath = path.join(prefix, entry.name);
      
      if (entry.isDirectory()) {
        const subFiles = await this.getFileList(
          path.join(dir, entry.name), 
          relativePath
        );
        files.push(...subFiles);
      } else {
        const stats = await fs.stat(path.join(dir, entry.name));
        files.push({
          path: relativePath,
          size: stats.size,
          modified: stats.mtime.toISOString()
        });
      }
    }

    return files;
  }
}

// CLI usage
if (import.meta.url === `file://${process.argv[1]}`) {
  const configFile = process.argv[2];
  const buildId = process.argv[3] || `build-${Date.now()}`;

  if (!configFile) {
    console.error('Usage: node build-tenant-site.js <config-file> [build-id]');
    process.exit(1);
  }

  try {
    const configContent = await fs.readFile(configFile, 'utf-8');
    const tenantConfig = JSON.parse(configContent);
    
    const builder = new TenantSiteBuilder();
    const result = await builder.buildSite(tenantConfig, buildId);
    
    if (result.success) {
      console.log('\n✅ Build completed successfully!');
      console.log(`📦 Deployment package: ${result.deploymentPackage.packagePath}`);
      process.exit(0);
    } else {
      console.error(`\n❌ Build failed: ${result.error}`);
      process.exit(1);
    }
  } catch (error) {
    console.error('❌ Error:', error.message);
    process.exit(1);
  }
}

export { TenantSiteBuilder };

File: tenant-site-template/build-tenant-site.py
#!/usr/bin/env python3
"""
Tenant Site Builder - Generates customized baby raffle sites
This script builds a tenant-specific site from the template
"""

import os
import json
import shutil
import subprocess
import sys
import argparse
from pathlib import Path
from typing import Dict, Any
import tempfile

def log_info(message: str):
    print(f"ℹ️  {message}")

def log_success(message: str):
    print(f"✅ {message}")

def log_error(message: str):
    print(f"❌ {message}")

def log_warning(message: str):
    print(f"⚠️  {message}")

def replace_placeholders(content: str, config: Dict[str, Any]) -> str:
    """Replace template placeholders with actual tenant configuration"""
    
    replacements = {
        '{{SUBDOMAIN}}': config.get('subdomain', ''),
        '{{SITE_NAME}}': config.get('site_name', ''),
        '{{PARENT_NAMES}}': config.get('parent_names', ''),
        '{{DUE_DATE}}': config.get('due_date', ''),
        '{{VENMO_ACCOUNT}}': config.get('venmo_account', ''),
        '{{PRIMARY_COLOR}}': config.get('primary_color', '#ec4899'),
        '{{SECONDARY_COLOR}}': config.get('secondary_color', '#8b5cf6'),
        '{{DESCRIPTION}}': config.get('description', ''),
        '{{API_BASE_URL}}': config.get('api_base_url', 'https://api.base2ml.com'),
        '{{TENANT_ID}}': config.get('tenant_id', ''),
        '{{SLIDESHOW_IMAGES}}': ', '.join([f'"{img}"' for img in config.get('slideshow_images', [])]),
        '{{LOGO_URL}}': config.get('logo_url', '')
    }
    
    result = content
    for placeholder, value in replacements.items():
        result = result.replace(placeholder, str(value))
    
    return result

def build_tenant_site(tenant_id: str, config: Dict[str, Any], output_dir: str) -> bool:
    """Build a customized tenant site"""
    
    try:
        log_info(f"Building site for tenant: {tenant_id}")
        log_info(f"Subdomain: {config.get('subdomain', 'unknown')}")
        
        # Create temporary build directory
        build_dir = tempfile.mkdtemp(prefix=f"tenant-{tenant_id}-")
        template_dir = Path(__file__).parent
        
        log_info(f"Using build directory: {build_dir}")
        
        # Copy template to build directory
        shutil.copytree(template_dir, f"{build_dir}/site", dirs_exist_ok=True)
        
        os.chdir(f"{build_dir}/site")
        
        # Process all template files
        log_info("Processing template files...")
        
        for root, dirs, files in os.walk("src"):
            for file in files:
                if file.endswith(('.tsx', '.ts', '.js', '.jsx', '.html', '.css')):
                    filepath = os.path.join(root, file)
                    
                    with open(filepath, 'r', encoding='utf-8') as f:
                        content = f.read()
                    
                    # Replace placeholders
                    content = replace_placeholders(content, config)
                    
                    with open(filepath, 'w', encoding='utf-8') as f:
                        f.write(content)
        
        # Update package.json with tenant-specific name
        package_json_path = "package.json"
        if os.path.exists(package_json_path):
            with open(package_json_path, 'r') as f:
                package_data = json.load(f)
            
            package_data['name'] = f"baby-raffle-{config.get('subdomain', tenant_id)}"
            
            with open(package_json_path, 'w') as f:
                json.dump(package_data, f, indent=2)
        
        # Create environment file for build
        env_content = f"""
VITE_API_BASE_URL={config.get('api_base_url', 'https://api.base2ml.com')}
VITE_TENANT_ID={tenant_id}
VITE_SUBDOMAIN={config.get('subdomain', '')}
"""
        
        with open('.env.production', 'w') as f:
            f.write(env_content.strip())
        
        # Install dependencies
        log_info("Installing dependencies...")
        result = subprocess.run(['npm', 'install'], capture_output=True, text=True)
        if result.returncode != 0:
            log_error(f"Failed to install dependencies: {result.stderr}")
            return False
        
        # Build the site
        log_info("Building production site...")
        result = subprocess.run(['npm', 'run', 'build'], capture_output=True, text=True)
        if result.returncode != 0:
            log_error(f"Build failed: {result.stderr}")
            return False
        
        # Copy build output to final destination
        if os.path.exists('dist'):
            os.makedirs(output_dir, exist_ok=True)
            shutil.copytree('dist', output_dir, dirs_exist_ok=True)
            log_success(f"Build output copied to: {output_dir}")
        else:
            log_error("Build output directory not found")
            return False
        
        # Cleanup
        shutil.rmtree(build_dir)
        
        log_success(f"Site built successfully for {config.get('subdomain')}.base2ml.com")
        return True
        
    except Exception as e:
        log_error(f"Build failed: {str(e)}")
        return False

def deploy_to_s3(subdomain: str, build_dir: str, bucket_name: str) -> bool:
    """Deploy built site to S3 bucket"""
    
    try:
        log_info(f"Deploying {subdomain} to S3...")
        
        # Create subdomain-specific S3 path
        s3_path = f"s3://{bucket_name}/sites/{subdomain}/"
        
        # Upload files
        result = subprocess.run([
            'aws', 's3', 'sync', build_dir, s3_path, 
            '--delete', '--cache-control', 'max-age=31536000'
        ], capture_output=True, text=True)
        
        if result.returncode != 0:
            log_error(f"S3 upload failed: {result.stderr}")
            return False
        
        log_success(f"Site deployed to S3: {s3_path}")
        return True
        
    except Exception as e:
        log_error(f"S3 deployment failed: {str(e)}")
        return False

def main():
    parser = argparse.ArgumentParser(description='Build and deploy tenant baby raffle site')
    parser.add_argument('--tenant-id', required=True, help='Tenant ID')
    parser.add_argument('--config-file', help='JSON config file path')
    parser.add_argument('--config-json', help='JSON config string')
    parser.add_argument('--output-dir', help='Output directory for build')
    parser.add_argument('--deploy-s3', help='S3 bucket name for deployment')
    parser.add_argument('--api-url', default='https://api.base2ml.com', help='API base URL')
    
    args = parser.parse_args()
    
    # Load configuration
    config = {}
    
    if args.config_file:
        with open(args.config_file, 'r') as f:
            config = json.load(f)
    elif args.config_json:
        config = json.loads(args.config_json)
    else:
        # Fetch from API
        import requests
        try:
            response = requests.get(f"{args.api_url}/api/site/config/{args.tenant_id}")
            response.raise_for_status()
            config = response.json()
        except Exception as e:
            log_error(f"Failed to fetch tenant config: {e}")
            sys.exit(1)
    
    # Set default output directory
    if not args.output_dir:
        args.output_dir = f"./builds/{config.get('subdomain', args.tenant_id)}"
    
    # Build the site
    success = build_tenant_site(args.tenant_id, config, args.output_dir)
    
    if not success:
        sys.exit(1)
    
    # Deploy to S3 if requested
    if args.deploy_s3:
        deploy_success = deploy_to_s3(
            config.get('subdomain', args.tenant_id), 
            args.output_dir, 
            args.deploy_s3
        )
        if not deploy_success:
            sys.exit(1)
    
    log_success("Tenant site build completed successfully!")

if __name__ == '__main__':
    main()